"""
SMS gRPC æœåŠ¡å™¨å®ç°
"""
import json
import time
import grpc
from concurrent import futures
from loguru import logger

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from src.common.modem_manager import ModemManager
from src.sms.sender import SMSSender
from src.sms import sms_pb2, sms_pb2_grpc


class SMSService(sms_pb2_grpc.SMSServiceServicer):
    """
    SMS gRPC æœåŠ¡å®ç°
    """

    def __init__(self, modem_manager: ModemManager, sender: SMSSender):
        self.modem_manager = modem_manager
        self.sender = sender

    async def SendSMS(self, request, context):
        """å‘é€å•æ¡çŸ­ä¿¡"""
        logger.info(f"ğŸ“¨ å‘é€çŸ­ä¿¡è¯·æ±‚: {request.phone_number}")

        try:
            # æ„å»ºå…ƒæ•°æ®
            metadata = dict(request.metadata)
            if request.sender_id:
                metadata['sender_id'] = request.sender_id
            metadata['delivery_report'] = str(request.delivery_report)

            # å‘é€çŸ­ä¿¡
            result = await self.sender.send(
                phone_number=request.phone_number,
                content=request.content,
                metadata=metadata
            )

            # æ„å»ºå“åº”
            status_code = 200 if result['success'] else 500

            return sms_pb2.SendSMSResponse(
                status=status_code,
                message=result['message'],
                data=json.dumps(result, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¤„ç†å‘é€çŸ­ä¿¡è¯·æ±‚å¤±è´¥: {e}")

            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "phone_number": request.phone_number,
                "success": False
            }

            return sms_pb2.SendSMSResponse(
                status=500,
                message=f"å†…éƒ¨æœåŠ¡å™¨é”™è¯¯: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def SendBatchSMS(self, request, context):
        """æ‰¹é‡å‘é€çŸ­ä¿¡"""
        logger.info(f"ğŸ“¦ æ‰¹é‡å‘é€çŸ­ä¿¡è¯·æ±‚ï¼Œæ•°é‡: {len(request.phone_numbers)}")

        try:
            # æ„å»ºå…ƒæ•°æ®
            metadata = dict(request.metadata)
            if request.sender_id:
                metadata['sender_id'] = request.sender_id
            metadata['delivery_report'] = str(request.delivery_report)

            # æ‰¹é‡å‘é€çŸ­ä¿¡
            result = await self.sender.send_batch(
                phone_numbers=list(request.phone_numbers),
                content=request.content,
                metadata=metadata
            )

            # æ„å»ºå“åº”
            overall_success = result['success_count'] > 0
            status_code = 200 if overall_success else 500

            return sms_pb2.SendBatchSMSResponse(
                status=status_code,
                message=f"æ‰¹é‡å‘é€å®Œæˆï¼ŒæˆåŠŸ {result['success_count']} æ¡ï¼Œå¤±è´¥ {result['failure_count']} æ¡",
                data=json.dumps(result, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¤„ç†æ‰¹é‡å‘é€è¯·æ±‚å¤±è´¥: {e}")

            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "phone_numbers_count": len(request.phone_numbers),
                "success": False
            }

            return sms_pb2.SendBatchSMSResponse(
                status=500,
                message=f"å†…éƒ¨æœåŠ¡å™¨é”™è¯¯: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def HealthCheck(self, request, context):
        """å¥åº·æ£€æŸ¥"""
        try:
            # æ£€æŸ¥è°ƒåˆ¶è§£è°ƒå™¨ç®¡ç†å™¨çŠ¶æ€
            modem_status = await self.modem_manager.get_status()
            health_status = await self.modem_manager.health_check()

            health_data = {
                "timestamp": time.time(),
                "service_ready": health_status,
                "health_status": "healthy" if health_status else "unhealthy",
                "modem_status": modem_status,
                "details": {
                    "total_modems": modem_status["total_modems"],
                    "available_modems": modem_status["available_modems"],
                    "in_use_modems": modem_status["in_use_modems"],
                    "initialized": modem_status["initialized"]
                }
            }

            status_code = 200 if health_status else 503

            return sms_pb2.HealthCheckResponse(
                status=status_code,
                message="æœåŠ¡å¥åº·" if health_status else "æœåŠ¡ä¸å¥åº·",
                data=json.dumps(health_data, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¥åº·æ£€æŸ¥å¤±è´¥: {e}")

            error_data = {
                "timestamp": time.time(),
                "service_ready": False,
                "error": str(e),
                "details": "å¥åº·æ£€æŸ¥å¼‚å¸¸"
            }

            return sms_pb2.HealthCheckResponse(
                status=500,
                message=f"å¥åº·æ£€æŸ¥å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def GetModemStatus(self, request, context):
        """è·å–è°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€"""
        try:
            status = await self.modem_manager.get_status()

            return sms_pb2.ModemStatusResponse(
                status=200,
                message="è°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€è·å–æˆåŠŸ",
                data=json.dumps(status, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ è·å–è°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€å¤±è´¥: {e}")

            error_data = {
                "timestamp": time.time(),
                "error": str(e),
                "details": "è·å–è°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€å¤±è´¥"
            }

            return sms_pb2.ModemStatusResponse(
                status=500,
                message=f"è·å–è°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )


def create_server(modem_manager: ModemManager, sender: SMSSender, max_workers: int = 10) -> grpc.aio.Server:
    """
    åˆ›å»º gRPC æœåŠ¡å™¨

    Args:
        modem_manager: è°ƒåˆ¶è§£è°ƒå™¨ç®¡ç†å™¨
        sender: çŸ­ä¿¡å‘é€å™¨
        max_workers: æœ€å¤§å·¥ä½œçº¿ç¨‹æ•°

    Returns:
        gRPC æœåŠ¡å™¨å®ä¾‹
    """
    server = grpc.aio.server(
        futures.ThreadPoolExecutor(max_workers=max_workers)
    )

    # æ·»åŠ æœåŠ¡
    sms_service = SMSService(modem_manager, sender)
    sms_pb2_grpc.add_SMSServiceServicer_to_server(sms_service, server)

    return server
